### ISSUE TITLE
Tlmgr Update error with a-Shell on Ipad Pro

### ISSUE THREAD

**Issue by Erd2024** [holzschu/a-shell#747]
I need to install a Package for texlive. When I try to update tlmgr, I get following error:

First directive needs to be 'name', not Wasm error: WASI Exit error: 0 at /private/var/containers/Bundle
/Application/9719B6CF-1435-4289-850B-01933FB443E4/a-Shell.app/Perl/TeXLive/TLPOBJ.pm line 106, <$retfh> 
line 375840.

I installed a-Shell new, but get the same error. Same, when I try to install the missing Package.

**Comment by holzschu** [comment]
Hi, 
thanks for reporting this issue. I can reproduce it, even with `tlmgr list`. I am now investigating to see if it's something I introduced recently. I assume you are using the latest version (1.12.8)?

**Comment by Erd2024** [comment]
Hi,

thank you for your quick response. yes, I’m using the latest Version.

Von: Nicolas Holzschuch ***@***.***>
Gesendet: Mittwoch, 31. Januar 2024 07:48
An: holzschu/a-shell ***@***.***>
Cc: Erdmann, Wolfgang ***@***.***>; Author ***@***.***>
Betreff: Re: [holzschu/a-shell] Tlmgr Update error with a-Shell on Ipad Pro (Issue #747)

Sie erhalten nicht oft eine E-Mail von ***@***.*** Erfahren Sie, warum dies wichtig ist<https://aka.ms/LearnAboutSenderIdentification>

Hi,
thanks for reporting this issue. I can reproduce it, even with tlmgr list. I am now investigating to see if it's something I introduced recently. I assume you are using the latest version (1.12.8)?

**Comment by holzschu** [comment]
I've found the issue; I'm making a new release, it will be fixed in version 1.12.9.

**Comment by holzschu** [comment]
Hi, version 1.12.9 is now released on the AppStore, with the fix in place.

**Comment by Erd2024** [comment]
Thank you very much for the information and the quick correction of the error. Thanks to a-shell I can edit my LaTeX texts on the go 😊

Kind regards,

Wolfgang

Von: Nicolas Holzschuch ***@***.***>
Gesendet: Freitag, 2. Februar 2024 08:04
An: holzschu/a-shell ***@***.***>
Cc: Erdmann, Wolfgang ***@***.***>; Author ***@***.***>
Betreff: Re: [holzschu/a-shell] Tlmgr Update error with a-Shell on Ipad Pro (Issue #747)

Sie erhalten nicht oft eine E-Mail von ***@***.*** Erfahren Sie, warum dies wichtig ist<https://aka.ms/LearnAboutSenderIdentification>

Hi, version 1.12.9 is now released on the AppStore, with the fix in place.

**Comment by Erd2024** [comment]
I have downloaded and installed the new version, but tlmgr is still not working:

Could not determine directory of tlmgr executable, maybe shared library woes?
Check for error messages above at /private/var/containers/Bundle/Application/85309C28-A9E3-422F-B389-8D0
7E6827480/a-Shell.app/bin/tlmgr line 72.
BEGIN failed--compilation aborted at /private/var/containers/Bundle/Application/85309C28-A9E3-422F-B389-
8D07E6827480/a-Shell.app/bin/tlmgr line 85.

**Comment by holzschu** [comment]
Hi,
this time I cannot reproduce the issue. tlmgr works, both with my development version and with the appstore version. 

I have no idea what is happening, even after looking at the lines of code around the error message, I would suspect a failed install or some environment variables set to the wrong value. 

Out of curiosity, what does "kpsewhich -var-value=TEXMFROOT" say?

**Comment by Erd2024** [comment]
Hi,

the output of kpsewhich -var-value=TEXMFROOT is:

warning: kpathsea: configuration file texmf.cnf not found in these directories: [long list of directories]

segmentation fault


Von: Nicolas Holzschuch ***@***.***>
Gesendet: Freitag, 2. Februar 2024 08:53
An: holzschu/a-shell ***@***.***>
Cc: Erdmann, Wolfgang ***@***.***>; Author ***@***.***>
Betreff: Re: [holzschu/a-shell] Tlmgr Update error with a-Shell on Ipad Pro (Issue #747)

Sie erhalten nicht oft eine E-Mail von ***@***.*** Erfahren Sie, warum dies wichtig ist<https://aka.ms/LearnAboutSenderIdentification>

Hi,
this time I cannot reproduce the issue. tlmgr works, both with my development version and with the appstore version.

I have no idea what is happening, even after looking at the lines of code around the error message, I would suspect a failed install or some environment variables set to the wrong value.

Out of curiosity, what does "kpsewhich -var-value=TEXMFROOT" say?

**Comment by holzschu** [comment]
This looks like something went wrong during the reinstall of the TeX files. If you had lots of TeX packages installed, I would suggest making a copy of ~/Library/texlive/2023/texmf-dist, then re-download TeX files: exit the app, deactivate TeX in settings, wait for a while, reopen the app, type tex and re-download TeX files.

(if you don't have many TeX packages installed, you can skip the copy part).

**Comment by Erd2024** [comment]
It was a local problem - I installed tex again and now tlmgr works.

**Comment by Erd2024** [comment]
Thank you for your help, I reinstalled tex and now tlmgr works.

Am 02.02.2024 um 09:56 schrieb Nicolas Holzschuch ***@***.***>:

﻿
Sie erhalten nicht oft eine E-Mail von ***@***.*** Erfahren Sie, warum dies wichtig ist<https://aka.ms/LearnAboutSenderIdentification>

This looks like something went wrong during the reinstall of the TeX files. If you had lots of TeX packages installed, I would suggest making a copy of ~/Library/texlive/2023/texmf-dist, then re-download TeX files: exit the app, deactivate TeX in settings, wait for a while, reopen the app, type tex and re-download TeX files.

(if you don't have many TeX packages installed, you can skip the copy part).

**Comment by holzschu** [comment]
That's good to hear. I would have preferred for it to work without the reinstall, but I'll take it.

(The mechanism to download TeX files goes through Apple servers (using something they call On-Demand-Resources). It's not fully reliable, and that's annoying, but I don't see many alternatives to replace it. I do not know of many places who will gladly host 1GB+ of data for reliable high-speed downloads for free).

**Comment by Erd2024** [comment]
I still have problems updating Latex with tlmgr. Tlmgr —list works. Maybe it is because tlmgr has the wrong Version:

[~/Documents]$ tlmgr --version
tlmgr revision 59291 (2021-05-21 05:14:40 +0200)
tlmgr using installation: /private/var/mobile/Containers/Data/Application/EE321E92-B219-4614-BFCA-7965B2
046EB8/Library/texlive/2023
TeX Live (https://tug.org/texlive) version 2023

I tried to update tlmgr (tlmgr update —self and downloading and running update-tlmgr-latent.sh), but it remains the 2021version.

Tlmgr update —all starts, but ends with errors. I tried several times, changed the repository, but did not succed in updating texlive.

**Comment by holzschu** [comment]
Hi, 
I'm also seeing weird issues with tlmgr. It complains about files not existing, even though these files do exist (I checked). These errors are not related to the texlive update (i.e. they persist after I use the tlmgr version 2024). 

I know that there is a hard limit for tlmgr at around 300 packages. That's because iOS is running out of file descriptors (it does not let an application allocate more than 1280 file descriptors, each time tlmgr calls curl to download, it opens 3 files descriptors, and I have problems with closing some of these). For this reason, if you run `tlmgr update --all`, it will start showing download errors after 300 repositories. 

Short version: 
- "file does not exist" errors: these seem to be related to when tlmgr generates the backup of the package. Annoying but harmless.
- "could not connect" errors: force-kill the app and restart tlmgr for the next 300 packages.

**Comment by Erd2024** [comment]
When Running ˋtlmgr update —allˋ I got errors after about 250 packages. I don‘t know how to stop tlmgr while running. Ctrl+c didn‘t work and I have no home-Button on my IPad Pro to force-kill the process. Tlmgr stopped running with errors and after that didn‘t start again:

/private/var/containers/Bundle/Application/D0428CBF-0EB6-4A98-B693-137227446890/a-Shell.app/bin/tlmgr: T
LPDB::from_file could not initialize from: https://mirror.ctan.org/systems/texlive/tlnet/tlpkg/texlive.t
lpdb
/private/var/containers/Bundle/Application/D0428CBF-0EB6-4A98-B693-137227446890/a-Shell.app/bin/tlmgr: M
aybe the repository setting should be changed.
/private/var/containers/Bundle/Application/D0428CBF-0EB6-4A98-B693-137227446890/a-Shell.app/bin/tlmgr: M
ore info: https://tug.org/texlive/acquire.html
